app/src/main/java/org/simple/clinic/editpatient/PatientEditScreenController.kt

```diff
@@ -0,0 +1,70 @@
+package org.simple.clinic.editpatient
+
+import io.reactivex.Observable
+import io.reactivex.ObservableSource
+import io.reactivex.ObservableTransformer
+import io.reactivex.rxkotlin.Observables
+import io.reactivex.rxkotlin.ofType
+import org.simple.clinic.ReportAnalyticsEvents
+import org.simple.clinic.patient.PatientRepository
+import org.simple.clinic.util.Just
+import org.simple.clinic.widgets.UiEvent
+import javax.inject.Inject
+
+typealias Ui = PatientEditScreen
+typealias UiChange = (Ui) -> Unit
+
+class PatientEditScreenController @Inject constructor(
+    private val patientRepository: PatientRepository
+) : ObservableTransformer<UiEvent, UiChange> {
+
+  override fun apply(events: Observable<UiEvent>): ObservableSource<UiChange> {
+    val replayedEvents = events.compose(ReportAnalyticsEvents()).replay().refCount()
+
+    return prefillOnStart(replayedEvents)
+  }
+
+  private fun prefillOnStart(events: Observable<UiEvent>): Observable<UiChange> {
+
+    val patientUuidStream = events.ofType<PatientEditScreenCreated>()
+        .map { it.patientUuid }
+
+    val savedPatient = patientUuidStream
+        .flatMap(patientRepository::patient)
+        .take(1)
+        .map { it as Just }
+        .map { it.value }
```

Comment:
Unwrapping Optionals in Rx chains looks so bad. One simpler way is to deconstruct:

```kotlin
.map { (patient) -> patient!! }
```

Can we create an extension function to simplify this somehow?

Reply:
I propose:

```kotlin
private fun <T : Any> Observable<Optional<T>>.filterJust(): Observable<T> {
  return map { it as Just<T> }.map { (value) -> value }
}

private fun <T : Any> Observable<Optional<T>>.mapAsJust(): Observable<T> {
  return filter { it is Just }.map { (value) -> value!! }
}
```

Not very satisfied with the naming though.

Reply:
I made a class `RxOptional.kt` since this is a pattern we seem to using in a lot of places. I've added one method for now which is basically

```kotlin
fun <T : Any> Observable<Optional<T>>.unwrapIfJust(): Observable<T> {
  return filter { it is Just }.map { (value) -> value!! }
}
```

Reply:
This looks much better. I noticed you're using `cache()` for sharing `savedPatient` stream. I don't have enough experience, but I think `Observable#cache()` is dangerous. From what I understand from the docs, there's some caveats:

```
<em>Note:</em> You sacrifice the ability to dispose the origin when you use the {@code cache} Observer so be careful not to use this Observer on ObservableSources that emit an infinite or very large number of items that will use up memory. A possible workaround is to apply `takeUntil` with a predicate or another source before (and perhaps after) the application of cache().
```

Let's use `replay().refCount()` instead?

As a rule of thumb, I use `cache()` with `Single` and `replay().refCount()` with `Observable`.

Reply:
As decided offline, I'll make this change as part of my next PR.